File: utils/Step2_Producer_Generation.R
################ part 2 ######################
# at the beginning of this part, the input io has 302 rows and 387 variables
#-----------------------------------------------------------------------------------
#Create producers/suppliers database
#-----------------------------------------------------------------------------------
# rm(list = ls())
# path2code <- '/Users/xiaodanxu/Documents/GitHub/SynthFirm/utils/'
# source(paste0(path2code, 'step0_SynthFirm_starter.R')) # load packages
#source(paste0(path2code, 'scenario/scenario_variables.R'))  # load environmental variable
#source('config.R')  # load input settings 

# path2file <-
#   "/Users/xiaodanxu/Documents/SynthFirm.nosync"
# setwd(path2file)
# try update file name
print("Generating synthetic producers...")

output_dir <- paste0(path2file, "/outputs_", out_scenario)

firms <- data.table::fread(paste0(output_dir, '/', synthetic_firms_no_location_file), h = T) # 8,396, 679 FIRMS
mesozone_faf_lookup <- data.table::fread(paste0(path2file, "/inputs_", scenario, '/', zonal_id_file), h = T)

c_n6_n6io_sctg <-
  data.table::fread(paste0(path2file, "/SynthFirm_parameters/", c_n6_n6io_sctg_file), h = T)
for_prod <-
  data.table::fread(paste0(path2file, "/SynthFirm_parameters/", foreign_prod_file), h = T)

io <- data.table::fread(paste0(path2file, "/SynthFirm_parameters/", BEA_io_2017_file), h = T)
unitcost <- data.table::fread(paste0(path2file, "/SynthFirm_parameters/", agg_unit_cost_file), h = T)

producer_value_fraction_by_location <- 
  data.table::fread(paste0(path2file, "/SynthFirm_parameters/", prod_by_zone_file), h = T)

sctg_lookup <- data.table::fread(paste0(path2file, "/SynthFirm_parameters/", SCTG_group_file), h = T)

# Creation of Producers database
# All agents that produce some SCTG commodity become potential producers
# wholesales are dealt with separately below
producers <-
  firms[Commodity_SCTG > 0 &
        substr(Industry_NAICS6_Make, 1, 2) != "42",] # 423,944 producers, exclude wholesale

# Create a table of production values and employment by NAICS code to calculate production value

#TODO change to use melt.data.table here to simplify
setnames(io, names(io), sub("X", "", names(io))) # Strip out the leading 'X' in the column names
io <-
  data.table(melt(io, id.vars = "make")) # Melt into long format and turn back into a data.table 
# 140,998 rows
setnames(
  io,
  c("make", "variable", "value"),
  c("Industry_NAICS6_Make", "Industry_NAICS6_Use", "ProVal") #make = from, use = to, use similar definition below
)
io[, Industry_NAICS6_Use := as.character(Industry_NAICS6_Use)]
io <- io[io$ProVal>0,]
write.csv(io, paste0(path2file, "/SynthFirm_parameters/", io_summary_file), row.names=FALSE) 
# write cleaned I-O data for future use


to_wholesale <- # commodity flow goes TO wholesale
  io[substr(Industry_NAICS6_Use, 1, 2) == "42" &
       ProVal > 0 ] # 1383 rows, 3 variables

naics_by_sctg_fration <- unique(c_n6_n6io_sctg[, list(SCTG = Commodity_SCTG, Industry_NAICS6_Make, Proportion)])
naics_by_sctg_fration <-
  naics_by_sctg_fration[SCTG > 0,]
#proportion of commodity by maker's industry
# wholesale_by_sctg <-
#   naics_by_sctg_fration[substr(Industry_NAICS6_Make, 1, 2) == "42" & SCTG > 0] # size = 395 * 4

naics_by_sctg_fration[, Proportion := Proportion / sum(Proportion), by = Industry_NAICS6_Make] 
# Re-distribution total production by wholesale segment and commodity

to_wholsale_by_sctg <-
  merge(naics_by_sctg_fration, to_wholesale, by = "Industry_NAICS6_Make") # 954 * 5
to_wholsale_by_sctg[, ProVal := ProVal * Proportion] 


# at this point, 'to_wholsale_by_sctg' refers to the amount of commodity from various industry TO wholesale, size = 954 * 5

from_wholesale <- io[substr(Industry_NAICS6_Make, 1, 2) == "42" &
                       ProVal > 0] # 859 rows
setnames(from_wholesale, "ProVal", "ProValFromWhl")


# at this point, 'from_wholesale_to_user' refers to the amount of commodity from wholesale to each end users (ProValFromWhl)

#which of those commodities can actually be sourced from wholesalers?
#TODO improve the naming here to be more explicit (e.g., fromwhl not a good name)
wholesale_flow <-
  merge(from_wholesale, to_wholsale_by_sctg, 
        by.x = 'Industry_NAICS6_Make', 
        by.y = 'Industry_NAICS6_Use', allow.cartesian = TRUE) #size = 81048 * 8

# at this point, fromwhl defines amount of commodity from wholesale and to wholesale

wholesale_flow[, ProValPctUse := ProVal / sum(ProVal), by = c('Industry_NAICS6_Make', 'Industry_NAICS6_Use')] 
# split among source production industry (to wholesale)

#allocate out the ProValFromWhl by weighting both input to wholesalers
#first need to scale production amount to consumption

wholesale_flow[, CellValue := ProValFromWhl * ProValPctUse]
whlprod <-
  sum(unique(wholesale_flow[, list(Industry_NAICS6_Make, ProVal)])$ProVal)
whlcons <-
  sum(wholesale_flow$CellValue)
wholesale_flow[, CellValue := CellValue * whlprod / whlcons]

#matrix of producers to consumers
#row totals are production amounts
#column totals are consumption amounts
#initial cell values are the proportion of consumption that comes from each producers by industry
#Then need to IPF to adjust cell values such that row and column totals are conserved


# for (i in 1:10) { # IPF method, col sum matches total wholesale production from manufacturers, col sum matches total consumption to end users
# 
#   wholesale_flow[, ColTotal := sum(CellValue), by = c('Industry_NAICS6_Make', 'Industry_NAICS6_Make.y')]
#   wholesale_flow[, ColWeight := ProVal / ColTotal]
#   wholesale_flow[, CellValue := CellValue * ColWeight]
#   wholesale_flow[, RowTotal := sum(CellValue), by = c('Industry_NAICS6_Make', 'Industry_NAICS6_Use')]
#   wholesale_flow[, RowWeight :=  ProValFromWhl/ RowTotal]
#   wholesale_flow[, CellValue := CellValue * RowWeight]
# }

wholesale_flow[, CellValue := round(CellValue)]
wholesale_flow <-
  wholesale_flow[CellValue > 0, list(Industry_NAICS6_Make = Industry_NAICS6_Make.y,
                                     Industry_NAICS6_Use,
                                     SCTG,
                                     NAICS_whl = Industry_NAICS6_Make,
                                     ProValWhl = CellValue)] #28747 * 5 after removing zero transactions
iowhl <-
  wholesale_flow[, list(ProValWhl = sum(ProValWhl)), by = list(Industry_NAICS6_Make, Industry_NAICS6_Use)]

#remove wholesale records from io table
io_no_wholesale <-
  io[substr(Industry_NAICS6_Use, 1, 2) != "42" &
       substr(Industry_NAICS6_Make, 1, 2) != "42"]

#replace wholesale records by adding the make-use value back to the io table
#as if it was direct from producer to consumer and not via whl
#so that the production amounts and consumption amounts will be correct
io_no_wholesale <-
  merge(
    io_no_wholesale,
    iowhl,
    by = c("Industry_NAICS6_Make", "Industry_NAICS6_Use"),
    all.x = TRUE
  )
io_no_wholesale[is.na(ProValWhl), ProValWhl := 0]
io_no_wholesale[, ProVal := ProVal + ProValWhl]
io_no_wholesale[, ProValWhl := NULL]

#add the wholesales with the correct capacities in value and tons
#to both producer and consumer tables
wholesalers <- firms[substr(Industry_NAICS6_Make, 1, 2) == "42",] # 407,124 wholesaler simulated
whlval <- wholesale_flow[, list(ProVal = sum(ProValWhl)), by = c('NAICS_whl', 'SCTG')]
setnames(whlval, "NAICS_whl", "Industry_NAICS6_Make")
setnames(whlval, "SCTG", "Commodity_SCTG")

list_of_wholesaler_naics <- unique(whlval$Industry_NAICS6_Make)
wholesalers[, Commodity_SCTG := NULL]
wholesalers_with_sctg <- NULL
for (whl_naics in list_of_wholesaler_naics){
  # print(whl_naics)
  con_firms <- wholesalers[Industry_NAICS6_Make == whl_naics,]
  sample_size <- nrow(con_firms)
  con_io <- whlval[Industry_NAICS6_Make == whl_naics,]
  con_io[, probability := ProVal/sum(ProVal)]
  sample_naics_make = sample_n(con_io, size = sample_size, weight = probability, replace = TRUE) %>% as_tibble()
  sample_naics_make <- as.data.table(sample_naics_make)
  con_firms <- cbind(con_firms, sample_naics_make[, 'Commodity_SCTG'])
  wholesalers_with_sctg <- rbind(wholesalers_with_sctg, con_firms)
  # break
} #301,825 wholesalers with sctg

wholesale_emp <- wholesalers_with_sctg[, list(Emp = sum(Emp)), by = c('Industry_NAICS6_Make', 'Commodity_SCTG', 'FAFZONE')] #5135

whlval_with_loc <-
  merge(wholesale_emp, whlval, 
  by = c("Industry_NAICS6_Make", 'Commodity_SCTG'))
whlval_with_loc <- merge(whlval_with_loc, 
                         producer_value_fraction_by_location[, list(Commodity_SCTG, FAFZONE = FAF, value_fraction)],
                         by = c("Commodity_SCTG", "FAFZONE")) # 31,656
whlval_with_loc[, value_fraction := value_fraction / sum(value_fraction), by = c('Industry_NAICS6_Make', 'Commodity_SCTG')] #production value per employee
whlval_with_loc[, ProVal := ProVal * value_fraction]
whlval_with_loc[, ValEmp := ProVal / Emp] #production value per employee
#wholesalers[, Commodity_SCTG := NULL]
wholesalers_with_value <-
  merge(wholesalers_with_sctg, whlval_with_loc[, list(Industry_NAICS6_Make, ValEmp, Commodity_SCTG, FAFZONE)], 
        by = c('Industry_NAICS6_Make', "Commodity_SCTG", 'FAFZONE')) #merge the value per employee back on to businesses

wholesalers_with_value[, ProdVal := Emp * ValEmp] #calculate production value for each establishment, 
# some firms are removed as there is no I-O data + SCTG information for that NAICS code
# 299,526 wholesaler, some were dropped as no matching production value found


#======= this markes as the end of wholesaler generation ============#

#issues - need to be marked as a wholesales using the NAICS code
#but need to be tagged with the correct make/use commodity seperate from their NAICS code
#this should be easy for the consumer side, check possible for the producers side
producer_emp = producers[, list(Emp = sum(Emp)), by = c('Industry_NAICS6_Make', 'Commodity_SCTG', 'FAFZONE')]

prodval <-
  merge(producer_emp, io_no_wholesale[, list(ProVal = sum(ProVal)), by = Industry_NAICS6_Make], "Industry_NAICS6_Make")

prodval_with_loc <- merge(prodval, 
                          producer_value_fraction_by_location[, list(Commodity_SCTG, FAFZONE = FAF, value_fraction)],
                          by = c("Commodity_SCTG", "FAFZONE")) # 31,656
prodval_with_loc[, value_fraction := value_fraction / sum(value_fraction), by = c('Industry_NAICS6_Make', 'Commodity_SCTG')] #production value per employee
prodval_with_loc[, ProVal := ProVal * value_fraction]
prodval_with_loc[, ValEmp := ProVal / Emp] #production value per employee (in Million of Dollars)
# write.csv(prodval, 'production_value_per_emp.csv')
producers <-
  merge(producers, prodval_with_loc[, list(Industry_NAICS6_Make, ValEmp, Commodity_SCTG, FAFZONE)], 
        by = c('Industry_NAICS6_Make', 'Commodity_SCTG', 'FAFZONE')) #394,389 producers

# NAICS 331314 (Secondary Smelting and Alloying of Aluminum) is excluded in this process as it is not in I/O table
producers[, ProdVal := Emp * ValEmp] #calculate production value for each establishment (in Million of Dollars)



################ part 2 END######################


################ part 3 ######################
# Add foreign producers - one agent per country per commodity
# add in the NAICS_Make code, group, and calculate employment requirements to support that production
prodval <-
  merge(io[, list(ProVal = sum(ProVal)), by = Industry_NAICS6_Make], 
        producers[, list(Emp = sum(Emp)), by = Industry_NAICS6_Make], "Industry_NAICS6_Make")
prodval[, ValEmp := ProVal / Emp] #production value per employee (in Million of Dollars)

setnames(for_prod, "Commodity_NAICS6", "Industry_NAICS6_CBP") # 28470 * 6 
for_prod <-
  merge(for_prod, c_n6_n6io_sctg[, list(Industry_NAICS6_CBP, Industry_NAICS6_Make, Commodity_SCTG)], by =
          "Industry_NAICS6_CBP") #Merge in the I/O NAICS codes and SCTG codes, size = 27786 * 8
for_prod <-
  for_prod[Commodity_SCTG > 0, list(ProdVal = sum(USImpVal) / 1000000), by =
             list(Industry_NAICS6_Make, CBPZONE, FAFZONE, Commodity_SCTG)] # 20,416 foreign forms that provide commodities in scope
for_prod <-
  merge(for_prod, prodval[, list(Industry_NAICS6_Make, ValEmp)], by = "Industry_NAICS6_Make", all.x =
          TRUE) # no firm loss in this step, good
#TODO check on these missing commodities/commodites without value in the IO table -- should there be US production too? should it have value?
for_prod <- for_prod[!is.na(ValEmp) & ValEmp != 0] # 15312 firms remaining
#estimate employment and size category
#update ValEmp using foreign producer adjustment

for_prod[, ValEmp := ValEmp * foreignprodcostfactor] #same adjustment applied to unitcost, so assumption is that quantity per employee is the same as domestic production
for_prod[, Emp := round(ProdVal / ValEmp)]
for_prod[, esizecat := findInterval(Emp, c(0, 20, 100, 250, 500, 1000, 2500, 5000))]
#producers' output in POUNDS (units costs converted to pounds)
unitcost[, UnitCost := UnitCost / 2000]
for_prod <- merge(for_prod, unitcost, "Commodity_SCTG")
#update unit cost using foreign producer adjustment
for_prod[, UnitCost := UnitCost * foreignprodcostfactor]
for_prod[, ProdCap := ProdVal * 1000000 / UnitCost] # ProdVal was in $M, ProdCap in pound
### -------------------------------------------------------------------------------------
## -- Heither, 03-11-2016: enumerate large foreign consumers into multiple firms (threshold reduced)
for_prod[ProdCap <= 500000000, est := 1] # max production capability as 500 million pound
for_prod[ProdCap > 500000000, est := ceiling(ProdCap / 500000000)]		## number of fims to create
for_prod[est > 1, ProdVal := ProdVal / est]								## update ProdVal for multiple firms
for_prod[est > 1, ProdCap := ProdCap / est]								## update ProdCap for multiple firms
for_prod <-
  for_prod[rep(seq_len(for_prod[, .N]), est),]				## Enumerates the foreign producers using the est variable.
for_prod[, est := NULL]

for_prod[, MESOZONE := CBPZONE + 30000L]
for_prod[, BusID := max(firms$BusID) + .I] # size = 20446 * 11
### -------------------------------------------------------------------------------------
# calculate other fields required in producers tables

# Calculate producers' output in POUNDS (units costs converted to pounds)
producers <- merge(producers, unitcost, "Commodity_SCTG")
producers[, ProdCap := ProdVal * 1000000 / UnitCost] # ProdVal was in $M
# Calculate wholesalers output in POUNDS
wholesalers_with_value <- merge(wholesalers_with_value, unitcost, "Commodity_SCTG")
# factor up unitcost to reflect wholesalers margin
wholesalers_with_value[, UnitCost := UnitCost * wholesalecostfactor]
wholesalers_with_value[, ProdCap := ProdVal * 1000000 / UnitCost] # ProdVal was in $M

# combine domestic and foreign producers
#producers[, c("FAFID") := NULL] 
producers <- rbind(producers, for_prod, use.names = TRUE) # 441,376 PRODUCERS
rm(for_prod)

# Prepare for Writing out a producers file for each NAICS, with each firm represented by:
# SellerID (BusID)  Zone (MESOZONE)	NAICS (NAICS6_Make)	Size (Emp)	OutputCommodity (SCTG_Make)	OutputCapacityTons (ProdCap)	NonTransportUnitCost (UnitCost)
producers[, c("FAFZONE", "CBPZONE", "esizecat", "ProdVal", "ValEmp") := NULL]
setnames(
  producers,
  c(
    "BusID",
    "MESOZONE",
    "Industry_NAICS6_Make",
    "Emp",
    "ProdCap",
    "UnitCost"
  ),
  c(
    "SellerID",
    "Zone",
    "NAICS",
    "Size",
    "OutputCapacitylb",
    "NonTransportUnitCost"
  )
)
producers[, OutputCommodity := NAICS] # size = 414835 * 12

#Add in wholesalers to producers
wholesalers_with_value[, c("FAFZONE", "CBPZONE", "esizecat", "ProdVal", "ValEmp") :=
              NULL]
setnames(
  wholesalers_with_value,
  c(
    "BusID",
    "MESOZONE",
    "Industry_NAICS6_Make",
    "Emp",
    "ProdCap",
    "UnitCost"
  ),
  c(
    "SellerID",
    "Zone",
    "NAICS",
    "Size",
    "OutputCapacitylb",
    "NonTransportUnitCost"
  )
)
#simulate the single specific NAICS commodity that the wholesaler deals in to simplify
#(wholesale NAICS are one to many NAICS commodities)
#each wholesale firm is identified with a specific NAICS and SCTG
#need probabilities for the match with NAICS commodity
whlnaics <-
  wholesale_flow[, list(ProValWhl = sum(ProValWhl)), by = list(Industry_NAICS6_Make, SCTG, NAICS =
                                                                 NAICS_whl)]
whlnaics[, ProbProValWhl := ProValWhl / sum(ProValWhl), by = list(SCTG, NAICS)]
setkey(whlnaics, NAICS, SCTG)
whlnaics[, CumProValWhl := cumsum(ProbProValWhl), by = list(SCTG, NAICS)]
whlnaicscombs <- unique(whlnaics[, list(NAICS, SCTG)])
wholesalers_with_value[, temprand := runif(.N)]
for (i in 1:nrow(whlnaicscombs)) {
  whlnaicsi <-
    whlnaics[NAICS == whlnaicscombs$NAICS[i] &
               SCTG == whlnaicscombs$SCTG[i]]
  wholesalers_with_value[NAICS == whlnaicscombs$NAICS[i] &
                Commodity_SCTG == whlnaicscombs$SCTG[i],
              OutputCommodity := whlnaicsi$Industry_NAICS6_Make[1 + findInterval(temprand, whlnaicsi$CumProValWhl)]]
}  # for each wholesaler, find the only consumer industry within a list of potential consumers
#TODO - clean up correspondences to avoid no matches here
wholesalers_with_value[, temprand := NULL]
wholesalers_with_value <- wholesalers_with_value[!is.na(OutputCommodity)]
# wholesalers_with_value[, c("V1") := NULL] 
producers <- rbind(producers, wholesalers_with_value) # size = 690,436 * 8
setkey(producers, OutputCommodity)

print('Total number of producers:')
print(nrow(producers))

print('Total number of wholesalers (among producers):')
print(nrow(wholesalers_with_value))

write.csv(wholesalers_with_value, paste0(output_dir, '/', synthetic_wholesaler_file), row.names=FALSE)
write.csv(producers, paste0(output_dir, '/', synthetic_producer_file), row.names=FALSE)
write.csv(io_no_wholesale, paste0(output_dir, '/', io_filtered_file), row.names=FALSE)

sctg_lookup <- sctg_lookup %>% select(SCTG_Code, SCTG_Group, SCTG_Name) %>% as_tibble()
producers = producers %>% left_join(sctg_lookup, by = c("Commodity_SCTG" = "SCTG_Code")) %>% as_tibble()

for (i in 1:5) {
  print(paste0("Processing SCTG Group ", i))
  
  g1_prods = producers %>% filter(SCTG_Group == i) %>% select(
    SCTG_Group,
    Commodity_SCTG,
    SellerID,
    Zone,
    NAICS,
    OutputCommodity,
    OutputCapacitylb
  ) %>% as_tibble()
  
  
  fwrite(g1_prods, paste0(output_dir, '/', synthetic_producer_by_sctg_filehead, i, ".csv"))
}  

print('Producer generation is done!')
#writing out done below once sampling identified
################ part 3 END######################

File: utils/Step_2_Producer_Generation.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Mar  4 10:20:28 2024

@author: xiaodanxu
"""

import pandas as pd
import os
import numpy as np
from pandas import read_csv

print("Generating synthetic producers...")

########################################################
#### step 1 - configure environment and load inputs ####
########################################################


scenario_name = 'Seattle'
out_scenario_name = 'Seattle'
file_path = '/Users/xiaodanxu/Documents/SynthFirm.nosync'
parameter_dir = 'SynthFirm_parameters'
input_dir = 'inputs_' + scenario_name
output_dir = 'outputs_' + out_scenario_name

# specifications that has been defined before
c_n6_n6io_sctg_file = 'corresp_naics6_n6io_sctg_revised.csv'
synthetic_firms_no_location_file = "synthetic_firms.csv" 

# specifications that are new to this code
zonal_id_file = "zonal_id_lookup_final.csv" # zonal ID lookup table 
foreign_prod_file = "data_foreign_prod.csv" # total foreign production (pre-defined from prior studies)
BEA_io_2017_file = "data_2017io_revised_USE_value_added.csv" # final scaled BEA I-O use table
agg_unit_cost_file = "data_unitcost_cfs2017.csv" # unit cost by commodity from CFS 2017 (all zones combined)
prod_by_zone_file = "producer_value_fraction_by_faf.csv" # Total production value by FAF zone from CFS 2017
SCTG_group_file = "SCTG_Groups_revised.csv" # Commodity type to group lookup (pre-defined)

# load inputs
firms = read_csv(os.path.join(file_path, output_dir, synthetic_firms_no_location_file), low_memory=False) # 8,396, 679 FIRMS
mesozone_faf_lookup = read_csv(os.path.join(file_path, input_dir, zonal_id_file))
c_n6_n6io_sctg = read_csv(os.path.join(file_path, parameter_dir, c_n6_n6io_sctg_file))
for_prod = read_csv(os.path.join(file_path, parameter_dir, foreign_prod_file))

io = read_csv(os.path.join(file_path, parameter_dir, BEA_io_2017_file))
unitcost = read_csv(os.path.join(file_path, parameter_dir, agg_unit_cost_file))

producer_value_fraction_by_location = \
    read_csv(os.path.join(file_path, parameter_dir, prod_by_zone_file))

sctg_lookup = read_csv(os.path.join(file_path, parameter_dir, SCTG_group_file))

# define constant
foreignprodcostfactor = 0.9     # producer cost factor for foreign produers (applied to unit costs) 
wholesalecostfactor = 1.2     # markup factor for wholesalers (applied to unit costs)

# define output
io_summary_file = "io_summary_revised.csv" # i-o table in long format, without zero values
synthetic_wholesaler_file = "synthetic_wholesaler.csv" # synthetic wholesaler (serve as both buyer and supplier)
synthetic_producer_file = "synthetic_producers.csv" # synthetic producer
synthetic_producer_by_sctg_filehead = "prods_sctg" # file head for synthetic producer by sctg
io_filtered_file = "data_2017io_filtered.csv" # processed I-O table, after dropping wholesale transaction

# define result directory
result_dir = os.path.join(file_path, output_dir)



# <codecell>
########################################################
#### step 2 - Creation of Producers database ###########
########################################################

# All agents that produce some SCTG commodity become potential producers
# wholesales are generated separately below

producers = firms.loc[firms['Commodity_SCTG'] > 0] 
producers = producers.loc[producers['Industry_NAICS6_Make'].str[:2] != '42']
# 424,120 producers, exclude wholesale

# convert I-O to long table
io = pd.melt(io, id_vars=['make'], 
        var_name='Industry_NAICS6_Use', value_name='ProVal')

io = io.reset_index()

io = io.rename(columns = {'make': 'Industry_NAICS6_Make'})
io = io.drop(columns = 'index')
io.loc[:, 'Industry_NAICS6_Use'] = io.loc[:, 'Industry_NAICS6_Use'].str[1:7]

io = io.loc[io['ProVal']>0]
io.to_csv(os.path.join(file_path, parameter_dir, io_summary_file), index = False) 

 

# <codecell>
########################################################
#### step 3 - Creation of wholesale producers ##########
########################################################
 
# generate wholesale production
to_wholesale = io.loc[io['Industry_NAICS6_Use'].str[:2] == "42"] # 1383 rows, 3 variables

c_n6_n6io_sctg = c_n6_n6io_sctg.loc[c_n6_n6io_sctg['Commodity_SCTG'] > 0] # keep industries that produce commodity (not service)
naics_by_sctg_fration = \
    c_n6_n6io_sctg.drop_duplicates(subset = ['Commodity_SCTG', 'Industry_NAICS6_Make', 'Proportion'], keep = 'first')
 
# rescaling production fraction
naics_by_sctg_fration.loc[:, 'Proportion'] = \
    naics_by_sctg_fration.loc[:, 'Proportion'] / \
        naics_by_sctg_fration.groupby(['Industry_NAICS6_Make'])['Proportion'].transform('sum') 


# for selected wholesale industry, define the industry of commodity supplier

to_wholesale_by_sctg = \
    pd.merge(to_wholesale, 
             naics_by_sctg_fration, 
             on = "Industry_NAICS6_Make", how = 'inner') # 954 * 5

to_wholesale_by_sctg.loc[:, 'ProVal'] = \
    to_wholesale_by_sctg.loc[:, 'ProVal'] * to_wholesale_by_sctg.loc[:, 'Proportion']

to_wholesale_by_sctg = \
    to_wholesale_by_sctg[['Industry_NAICS6_Make', 'Industry_NAICS6_Use', 'ProVal', 'Commodity_SCTG', 'Proportion']]


# define wholesale industry as supplier

from_wholesale = io.loc[io['Industry_NAICS6_Make'].str[:2] == "42"] # 859 rows
from_wholesale = from_wholesale.rename(columns = {'ProVal': 'ProValFromWhl'})

wholesale_flow = \
  pd.merge(to_wholesale_by_sctg, from_wholesale, 
        left_on = 'Industry_NAICS6_Use', 
        right_on = 'Industry_NAICS6_Make', how = 'outer') #size = 82508 * 8
 # <codecell>  
  
whlprod =  from_wholesale.loc[:, 'ProValFromWhl'].sum() # this is the margin = sales - cost of good sold
whlcons =  to_wholesale_by_sctg.loc[:, 'ProVal'].sum()

wholesale_flow.loc[:, 'ProValPctUse'] = wholesale_flow.loc[:, 'ProVal']/ \
wholesale_flow.groupby(['Industry_NAICS6_Make_y', 'Industry_NAICS6_Use_y'])['ProVal'].transform('sum')

wholesalecostfactor = (whlcons + whlprod) / whlcons # assume zero inventory -> all commodity purchased will be sold

wholesale_flow.loc[:, 'CellValue'] = wholesale_flow.loc[:, 'ProValPctUse'] * wholesale_flow.loc[:, 'ProValFromWhl'] 

wholesale_flow.loc[:, 'CellValue'] *= whlcons/whlprod # scale value to cost of good
wholesale_flow.loc[:, 'CellValue'] *= wholesalecostfactor # add margin
wholesale_flow.loc[:, 'CellValue'] = np.round(wholesale_flow.loc[:, 'CellValue'], 0)
wholesale_flow = wholesale_flow.loc[wholesale_flow['CellValue'] > 0]
print(wholesale_flow.loc[:, 'CellValue'].sum())

wholesale_flow = \
    wholesale_flow[['Industry_NAICS6_Make_x', 'Industry_NAICS6_Use_y', 'Commodity_SCTG', 'Industry_NAICS6_Make_y', 'CellValue']]

wholesale_flow.columns =['Industry_NAICS6_Make', 'Industry_NAICS6_Use', 'SCTG', 'NAICS_whl', 'ProValWhl']    

# 21,336 rows


iowhl = wholesale_flow.groupby(['Industry_NAICS6_Make', 'Industry_NAICS6_Use'])[['ProValWhl']].sum()
iowhl = iowhl.reset_index()
print(iowhl.loc[:, 'ProValWhl'].sum())
# 10533 row

  
 
 #remove wholesale records from io table
io_no_wholesale = \
  io.loc[(io['Industry_NAICS6_Use'].str[:2] != "42") & (io['Industry_NAICS6_Make'].str[:2] != "42")]

io_with_wholesale = pd.merge(io_no_wholesale, iowhl,
                           on = ["Industry_NAICS6_Make", "Industry_NAICS6_Use"],
                           how = 'outer')
# some industry only transact with wholesaler, make sure those transactions are captured using outer join
io_with_wholesale.fillna(0, inplace=True)
io_with_wholesale.loc[:, 'ProVal_with_whl'] = \
    io_with_wholesale.loc[:, 'ProVal'] + io_with_wholesale.loc[:, 'ProValWhl'] 

 # <codecell>    
#add the wholesales with the correct capacities in value and tons
#to both producer and consumer tables
wholesalers = firms.loc[firms['Industry_NAICS6_Make'].str[:2] == "42",] 
# 413,532 wholesaler simulated

whlval = wholesale_flow.groupby(['NAICS_whl', 'SCTG'])[['ProValWhl']].sum()
whlval = whlval.reset_index()
whlval.columns = ['Industry_NAICS6_Make', 'Commodity_SCTG', 'ProVal']
# 270 rows
list_of_wholesaler_naics = whlval['Industry_NAICS6_Make'].unique()
wholesalers = wholesalers.drop(columns = ['Commodity_SCTG'])
wholesalers_with_sctg = None

for whl_naics in list_of_wholesaler_naics:
    print(whl_naics)
    con_firms = wholesalers.loc[wholesalers['Industry_NAICS6_Make'] == whl_naics]
    sample_size = len(con_firms)
    con_io = whlval.loc[whlval['Industry_NAICS6_Make'] == whl_naics]
    con_io.loc[:, 'probability'] = con_io.loc[:, 'ProVal'] / con_io.loc[:, 'ProVal'].sum()
    sample_naics_make = con_io.sample(n = sample_size, replace=True, weights = con_io['probability'],random_state=1)
    sample_naics_make = sample_naics_make.reset_index()
    con_firms = pd.concat([con_firms.reset_index(), sample_naics_make[['Commodity_SCTG']]], axis = 1)
    wholesalers_with_sctg = pd.concat([wholesalers_with_sctg, con_firms])

# 308235 wholesalers in the output that sell commodities

 # <codecell> 
 
# assign production for each wholesaler
wholesale_emp = \
    wholesalers_with_sctg.groupby(['Industry_NAICS6_Make', 'Commodity_SCTG', 'FAFZONE'])[['Emp']].sum()

wholesale_emp = wholesale_emp.reset_index()
whlval_with_loc = \
  pd.merge(wholesale_emp, whlval, 
  on = ["Industry_NAICS6_Make", 'Commodity_SCTG'], how = 'left')
  
producer_value_fraction_by_location = \
    producer_value_fraction_by_location.rename(columns = {'FAF':'FAFZONE'})
whlval_with_loc = \
    pd.merge(whlval_with_loc,
             producer_value_fraction_by_location[['Commodity_SCTG', 'FAFZONE', 'value_fraction']],
             on = ["Commodity_SCTG", "FAFZONE"], how = 'left') # 31,656

whlval_with_loc.loc[:, 'value_fraction'] = \
    whlval_with_loc.loc[:, 'value_fraction'] / \
        whlval_with_loc.groupby(['Industry_NAICS6_Make', 'Commodity_SCTG'])['value_fraction'].transform('sum')

whlval_with_loc.loc[:, 'ProVal'] = \
    whlval_with_loc.loc[:, 'ProVal'] * whlval_with_loc.loc[:, 'value_fraction'] 
whlval_with_loc.loc[:, 'ValEmp'] = \
    whlval_with_loc.loc[:, 'ProVal'] / whlval_with_loc.loc[:, 'Emp'] 

wholesalers_with_value = \
  pd.merge(wholesalers_with_sctg, 
           whlval_with_loc[['Industry_NAICS6_Make', 'ValEmp', 'Commodity_SCTG', 'FAFZONE']], 
           on = ['Industry_NAICS6_Make', "Commodity_SCTG", 'FAFZONE'],
           how = 'left') #merge the value per employee back on to businesses

wholesalers_with_value.loc[:, 'ProdVal'] = \
    wholesalers_with_value.loc[:, 'ValEmp']  * wholesalers_with_value.loc[:, 'Emp'] #calculate production value for each establishment, 

#======= this markes as the end of wholesaler generation ============#

#issues - need to be marked as a wholesales using the NAICS code
#but need to be tagged with the correct make/use commodity seperate from their NAICS code
#this should be easy for the consumer side, check possible for the producers side


# <codecell>
########################################################
#### step 4 - Creation of non-wholesale producers ######
########################################################

producer_emp = \
    producers.groupby(['Industry_NAICS6_Make', 'Commodity_SCTG', 'FAFZONE'])[['Emp']].sum()

producer_emp = producer_emp.reset_index()    

# 27085 * 4
